xds/src/test/java/io/grpc/xds/ClientXdsClientDataTest.java

```diff
@@ -773,245 +803,197 @@ public void parseCluster_ringHashLbPolicy_invalidRingSizeConfig_tooLargeRingSize
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail())
-        .isEqualTo("filerChain  has to have envoy.http_connection_manager");
+    thrown.expect(ResourceInvalidException.class);
+    thrown.expectMessage(
+        "FilterChain filter-chain-foo missing required HttpConnectionManager filter");
+    ClientXdsClient.parseServerSideListener(listener, null, true);
   }
 
   @Test
-  public void parseServerSideListener_duplicateFilterName() {
+  public void parseServerSideListener_duplicateFilter() throws ResourceInvalidException {
+    Filter filter = buildHttpConnectionManagerFilter(
+        HttpFilter.newBuilder().setName("http-filter-foo").setIsOptional(true).build());
     FilterChain filterChain =
-        buildFilterChain(
-            Filter.newBuilder()
-                .setName("envoy.http_connection_manager")
-                .setTypedConfig(Any.pack(HttpConnectionManager.getDefaultInstance()))
-                .build(),
-            Filter.newBuilder()
-                .setName("envoy.http_connection_manager")
-                .setTypedConfig(Any.pack(HttpConnectionManager.getDefaultInstance()))
-                .build());
+        buildFilterChain("filter-chain-foo", filter, filter);
     Listener listener =
         Listener.newBuilder()
             .setName("listener1")
             .setTrafficDirection(TrafficDirection.INBOUND)
             .addFilterChains(filterChain)
+            .setAddress(
+                Address.newBuilder()
+                    .setSocketAddress(
+                        SocketAddress.newBuilder().setAddress("10.0.1.2").setPortValue(8080)))
             .build();
-    StructOrError<io.grpc.xds.EnvoyServerProtoData.Listener> struct =
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail())
-        .isEqualTo("filerChain  has non-unique filter name:envoy.http_connection_manager");
+    thrown.expect(ResourceInvalidException.class);
+    thrown.expectMessage(
+        "FilterChain filter-chain-foo with duplicated filter: envoy.http_connection_manager");
+    ClientXdsClient.parseServerSideListener(listener, null, true);
   }
 
   @Test
-  public void parseServerSideListener_configDiscoveryFilter() {
+  public void parseServerSideListener_configDiscoveryFilter() throws ResourceInvalidException {
     Filter filter =
         Filter.newBuilder()
             .setName("envoy.http_connection_manager")
             .setConfigDiscovery(ExtensionConfigSource.newBuilder().build())
             .build();
-    FilterChain filterChain = buildFilterChain(filter);
+    FilterChain filterChain = buildFilterChain("filter-chain-foo", filter);
     Listener listener =
         Listener.newBuilder()
             .setName("listener1")
             .setTrafficDirection(TrafficDirection.INBOUND)
             .addFilterChains(filterChain)
+            .setAddress(
```

Comment:
Any reason why this was added? The test passes without setting the address.

Reply:
Okay, address is optional (`@Nullable` in the parsed form). Deleted.